src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2702,19 +2702,52 @@ private int LearnFromNullTest(BoundExpression expression, ref LocalState state)
             // We should not blindly strip conversions here. Tracked by https://github.com/dotnet/roslyn/issues/36164
             var expressionWithoutConversion = RemoveConversion(expression, includeExplicitConversions: true).expression;
             var slot = MakeSlot(expressionWithoutConversion);
-            return LearnFromNullTest(slot, expressionWithoutConversion.Type, ref state);
+
+            // Since we know for sure the slot is null (we just tested it), we know that dependent slots are not
+            // reachable and therefore can be treated as not null.  However, we have not computed the proper
+            // (inferred) type for the expression, so we cannot compute the correct symbols for the member slots here
+            // (using the incorrect symbols would result in computing an incorrect default state for them).
+            // Therefore we do not mark dependent slots not null.  See https://github.com/dotnet/roslyn/issues/39624
+            return LearnFromNullTest(slot, expressionWithoutConversion.Type, ref state, markDependentSlotsNotNull: false);
         }
 
-        private int LearnFromNullTest(int slot, TypeSymbol expressionType, ref LocalState state)
+        private int LearnFromNullTest(int slot, TypeSymbol expressionType, ref LocalState state, bool markDependentSlotsNotNull)
         {
             if (slot > 0 && PossiblyNullableType(expressionType))
             {
                 state[slot] = NullableFlowState.MaybeNull;
+                if (markDependentSlotsNotNull)
+                {
+                    MarkDependentSlotsNotNull(slot, expressionType, ref state);
+                }
             }
 
             return slot;
         }
 
+        // If we know for sure that a slot contains a null value, then we know for sure that dependent slots
+        // are "unreachable" so we might as well treat them as not null.  That way when this state is merged
+        // with another state, those dependent states won't pollute values from the other state.
+        private void MarkDependentSlotsNotNull(int slot, TypeSymbol expressionType, ref LocalState state)
+        {
+            foreach (var member in expressionType.GetMembers())
```

Comment:
This seems like a lot of potentially unnecessary work. Can we instead check at the point of use of a slot whether its root slot is null?

Reply:
Or could we loop over all the known slots rather than looping over members? See `InheritNullableStateOfTrackableType` for instance:
```C#
for (int s = slot + 1; s < nextVariableSlot; s++)
{
        var variable = variableBySlot[s];
        if (variable.ContainingSlot != valueSlot)
        {
            continue;
        }
        var member = variable.Symbol;
        // ...
}
```

---
In reply to: [357859900](https://github.com/dotnet/roslyn/pull/39625#discussion_r357859900) [](ancestors = 357859900)

Reply:
No, that doesn't work.  It might NOT be null, which is the case in the original bug report.  It is just that there was a null on a different branch with which we merged.  If it wasn't due to a pure null test, the null does not persist.

Reply:
No, we don't want to inherit the default state for slots not yet defined.

---
In reply to: [357865066](https://github.com/dotnet/roslyn/pull/39625#discussion_r357865066) [](ancestors = 357865066,357859900)